core/src/main/java/org/apache/iceberg/ReplaceManifests.java

```diff
@@ -0,0 +1,217 @@
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+
+import static org.apache.iceberg.TableProperties.MANIFEST_TARGET_SIZE_BYTES;
+import static org.apache.iceberg.TableProperties.MANIFEST_TARGET_SIZE_BYTES_DEFAULT;
+
+
+public class ReplaceManifests extends SnapshotProducer<RewriteManifests> implements RewriteManifests {
+  private final TableOperations ops;
+  private final PartitionSpec spec;
+  private final long manifestTargetSizeBytes;
+  private final SnapshotSummary.Builder summaryBuilder = SnapshotSummary.builder();
+  private final List<ManifestFile> newManifests = Collections.synchronizedList(Lists.newArrayList());
+  private final Map<Object, WriterWrapper> writers = Collections.synchronizedMap(new HashMap<>());
+  private final AtomicInteger manifestCount = new AtomicInteger(0);
+
+  private Function<DataFile, Object> clusterByFunc;
+  private Function<ManifestFile, Boolean> filterFunc;
+
+  ReplaceManifests(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.spec = ops.current().spec();
+    this.manifestTargetSizeBytes =
+      ops.current().propertyAsLong(MANIFEST_TARGET_SIZE_BYTES, MANIFEST_TARGET_SIZE_BYTES_DEFAULT);
+  }
+
+  @Override
+  protected String operation() {
+    return DataOperations.REPLACE;
+  }
+
+  @Override
+  public RewriteManifests set(String property, String value) {
+    summaryBuilder.set(property, value);
+    return this;
+  }
+
+  @Override
+  protected Map<String, String> summary() {
+    return summaryBuilder.build();
+  }
+
+  @Override
+  public ReplaceManifests clusterBy(Function<DataFile, Object> func) {
+    this.clusterByFunc = func;
+    return this;
+  }
+
+  @Override
+  public ReplaceManifests filter(Function<ManifestFile, Boolean> func) {
+    this.filterFunc = func;
+    return this;
+  }
+
+  @Override
+  public List<ManifestFile> apply(TableMetadata base) {
+    Preconditions.checkNotNull(clusterByFunc, "clusterBy function cannot be null");
+
+    List<ManifestFile> existingManifests = Collections.synchronizedList(new ArrayList<>());
+
+    try {
+      Tasks.foreach(base.currentSnapshot().manifests())
+          .executeWith(ThreadPools.getWorkerPool())
+          .run(manifest -> {
+            if (filterFunc != null && !filterFunc.apply(manifest)) {
+              existingManifests.add(manifest);
+            } else {
+              long entryNum = manifest.addedFilesCount() + manifest.existingFilesCount() + manifest.deletedFilesCount();
+              long avgEntryLen = manifest.length() / entryNum;
+
+              try (ManifestReader reader =
+                     ManifestReader.read(ops.io().newInputFile(manifest.path()))) {
+                FilteredManifest filteredManifest = reader.select(Lists.newArrayList("*"));
+                filteredManifest.iterator().forEachRemaining(
+                    file -> appendFile(file, avgEntryLen, clusterByFunc.apply(file))
```

Comment:
This needs to write `ManifestEntry` instead of `DataFile`. The public API for manifests returns `DataFile`, but internally we use `ManifestEntry` that encodes the snapshot when the file was added to the table. We need to preserve that.

Reply:
I changed this to use ManifestEntry and writer.addExisting()